primitives/io/src/tasks.rs

```diff
@@ -0,0 +1,237 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Runtime tasks.
+//!
+//! Contains runtime-usable functions for spawning parallel purely computational tasks.
+//!
+//! NOTE: This is experimental API.
+//! NOTE: When using in actual runtime, make sure you don't produce unbounded parallelism.
+//! So this is bad example to use it:
+//! ```rust
+//!    fn my_parallel_computator(data: Vec<u8>) -> Vec<u8> {
+//!        unimplemented!()
+//!    }
+//!    fn test(dynamic_variable: i32) {
+//!        for _ in 0..dynamic_variable { sp_io::tasks::spawn(my_parallel_computator, vec![]); }
+//!    }
+//! ```
+//!
+//! While this is a good example:
+//! ```rust
+//!    use codec::Encode;
+//!    static STATIC_VARIABLE: i32 = 4;
+//!
+//!    fn my_parallel_computator(data: Vec<u8>) -> Vec<u8> {
+//!        unimplemented!()
+//!    }
+//!
+//!    fn test(computation_payload: Vec<u8>) {
+//!        let parallel_tasks = (0..STATIC_VARIABLE).map(|idx|
+//!            sp_io::tasks::spawn(my_parallel_computator, computation_payload.chunks(10).nth(idx as _).encode())
+//!        );
+//!    }
+//! ```
+//!
+//! When allowing unbounded parallelism, malicious transactions can exploit it and partition
+//! network consensus based on how much resources nodes have.
+//!
+#[cfg(feature = "std")]
+mod inner {
+	use std::{panic::AssertUnwindSafe, sync::mpsc};
+	use sp_externalities::ExternalitiesExt as _;
+	use crate::TaskExecutorExt;
+
+	/// Task handle (wasm).
+	///
+	/// This can be `join`-ed to get (blocking) the result of
+	/// the spawned task execution.
+	#[must_use]
+	pub struct DataJoinHandle {
+		receiver: mpsc::Receiver<Vec<u8>>,
+	}
+
+	impl DataJoinHandle {
+		/// Join handle returned by `spawn` function
+		pub fn join(self) -> Vec<u8> {
+			self.receiver.recv().expect("Essential task failure: spawned runtime terminated before sending result.")
```

Comment:
I am not sure, why do you call it an essential task? I am not sure, but I had a preconception that `essential task` is such a task that if fails it has a license to bring down the whole process.

In any case, should we just leave: `spawned runtime terminated before sending result`?